# Food Health Analysis

KATE expects your code to define variables with specific names that correspond to certain things we are interested in.

KATE will run your notebook from top to bottom and check the latest value of those variables, so make sure you don't overwrite them.

* Remember to uncomment the line assigning the variable to your answer and don't change the variable or function names.
* Use copies of the original or previous DataFrames to make sure you do not overwrite them by mistake.

You will find instructions below about how to define each variable.

Once you're happy with your code, upload your notebook to KATE to check your feedback.

# How to complete and submit
Each exercise will look something like this:

```python
example_query = ''
#example_result = pd.read_sql(example_query, conn)
```

In each exercise you will need to define a query variable by writing the SQL code that you think will solve the problem. Once you have your query, uncomment the 2nd line, this will execute it and load the resulting data into a dataframe.

Nothing else needs to be changed in the 2nd line besides uncommenting it. 

After running this you will be free to inspect the result produced to see whether it's what you'd expect as the result. KATE will look for variables with the names defined in this notebook, so it is important not to rename the variables defined in this notebook.

Once you've completed the exercises upload this notebook to **KATE** to get feedback. You can also upload the notebook when you only have parts of it completed - if you do so, make sure you do not uncomment the `pd.read_sql` lines for which you don't have a query yet.

Refer to the instructions on **KATE** for more details on the dataset.

# Setup

The below code is setting up a connection to the SQLite Database. 

**Do not change this code!** The `conn` variable will be used throughout the notebook to query the database.

In [6]:
import pandas as pd
import sqlite3

conn = sqlite3.connect('data/sfscores.sqlite')

# Background

This dataset has been taken from an open data platform (found __[here](https://data.sfgov.org/)__) hosted by the city of San Francisco. It is an sqlite file which is a database created using SQLite. SQLite is a lightweight verion of SQL which features similar syntax to the vast majority of SQL flavours. It provides a fast way of accessing databases and storing databases locally. 

The data includes restaurant scores produced by the San Francisco health department up to and including the year 2016. These scores are generated by a health inspector who bases them upon the violations observed, these violations fall into three categories: High risk, Moderate risk or Low risk. 

The database itself consists of three tables: 

* _**businesses**_: information relating to restaurant businesses such as the owner and location
* _**inspections**_: information about individual inspection events
* _**violations**_: information about violation events



Referencing these tables and their respective columns will be useful in answering the following questions. The code below will show the column names (shown as name) and datatypes (shown as type) for the businesses table. To see the same information for the other tables try substituting violations or inspections for businesses in the first line of code.


In [7]:
# Code to show column names and data types within each table
table_info_query = 'PRAGMA table_info(businesses)'
table_info = pd.read_sql(table_info_query, conn)
table_info[['name','type']]


name         type
0            business_id      INTEGER
1                   name  VARCHAR(64)
2                address  VARCHAR(50)
3                   city  VARCHAR(23)
4            postal_code   VARCHAR(9)
5               latitude        FLOAT
6              longitude        FLOAT
7           phone_number       BIGINT
8                TaxCode   VARCHAR(4)
9   business_certificate      INTEGER
10      application_date         DATE
11            owner_name  VARCHAR(99)
12         owner_address  VARCHAR(74)
13            owner_city  VARCHAR(22)
14           owner_state  VARCHAR(14)
15             owner_zip  VARCHAR(15)

In [8]:
# Code to show column names and data types within each table
table_info_query = 'PRAGMA table_info(violations)'
table_info = pd.read_sql(table_info_query, conn)
table_info[['name','type']]


name     type
0      business_id     TEXT
1             date  INTEGER
2  ViolationTypeID     TEXT
3    risk_category     TEXT
4      description     TEXT

In [9]:
# Code to show column names and data types within each table
table_info_query = 'PRAGMA table_info(inspections)'
table_info = pd.read_sql(table_info_query, conn)
table_info[['name','type']]


name          type
0  business_id          TEXT
1        Score       INTEGER
2         date       INTEGER
3         type  VARCHAR (33)

# Queries

## Part 1: Essentials

**1. Write a SQL query that finds the number of business ids in the businesses table**

In [10]:
# Add your code below
number_of_businesses_query = '''
SELECT
    COUNT(*)
FROM businesses
'''
number_of_businesses_result = pd.read_sql(number_of_businesses_query, conn)
number_of_businesses_result

COUNT(*)
0      7616

**2. Write a SQL query that finds out how many unique business names are registered with San Francisco food health department (i.e. all uniques businesses in the businesses table) and name the column as unique restaurant name count.**

In [11]:
# Add your code below
unique_business_names_query = '''
SELECT
    COUNT(DISTINCT(name)) AS 'unique restaurant name count'
FROM businesses
'''
unique_business_names_result = pd.read_sql(unique_business_names_query, conn)
unique_business_names_result

unique restaurant name count
0                          7057

**3. Write a SQL query that finds out what is the earliest and latest date a health investigation is recorded in this database, you will find this information in the inspections table. Name these columns 'earliest date' and 'latest date'.**

In [12]:
# Add your code below
earliest_and_latest_investigation_query = '''
SELECT 
    MIN(date) AS 'earliest date',
    MAX(date) AS 'latest date'
FROM inspections
'''
earliest_and_latest_investigation_result = pd.read_sql(earliest_and_latest_investigation_query, conn)
earliest_and_latest_investigation_result

earliest date  latest date
0       20131007     20161215

**4. How many businesses are there in San Francisco where their owners live in the same area (postal code/zip code) as the business is located?**

In [13]:
# Add your code below
businesses_with_owners_nearby_query = '''
SELECT 
    COUNT(*)
FROM businesses
WHERE postal_code == owner_zip
'''
businesses_with_owners_nearby_result = pd.read_sql(businesses_with_owners_nearby_query, conn)
businesses_with_owners_nearby_result

COUNT(*)
0      4052

**5. Out of those businesses, how many of them has a registered business certificate?**

In [14]:
# Add your code below
businesses_with_registered_certificate_query = '''
SELECT 
    COUNT(*)
FROM businesses
WHERE postal_code == owner_zip
AND business_certificate != 'NaN'
'''
businesses_with_registered_certificate_result = pd.read_sql(businesses_with_registered_certificate_query, conn)
businesses_with_registered_certificate_result

COUNT(*)
0      3373

## Part 2: Groupby

**6. Find out the distribution of the risk exposure of all the violations reported in the database (i.e how many low, moderate and high risk violations are recorded). The first column of the result should be 'risk_category' and the second column the count which should be callled 'frequency'.**

In [15]:
# Add your code below
distribution_of_risk_exposure_query = '''
SELECT 
    risk_category,
    COUNT(risk_category) AS frequency
FROM violations
GROUP BY risk_category
'''
distribution_of_risk_exposure_result = pd.read_sql(distribution_of_risk_exposure_query, conn)
distribution_of_risk_exposure_result

risk_category  frequency
0      High Risk       5608
1       Low Risk      20996
2  Moderate Risk      14131

**7. Find out the distribution of the risk exposure of all the violations reported in the database that are *water related*. Sort them by frequency (count) from high to low.**

In [16]:
# Add your code below
distribution_of_water_risk_exposure_query = '''
SELECT
    risk_category,
    COUNT(risk_category) AS frequency
FROM 
    violations
WHERE 
    description LIKE '%water%'
GROUP BY 
    risk_category
ORDER BY 
    frequency DESC
'''
distribution_of_water_risk_exposure_result = pd.read_sql(distribution_of_water_risk_exposure_query, conn)
distribution_of_water_risk_exposure_result

risk_category  frequency
0  Moderate Risk        630
1      High Risk        237

**8. What types of inspections do the authorities conduct and how often do they occur in general. Calculate the distribution of different types of inspections with their frequency (type, frequency) based on inspections records. Sort them in ascending order based on frequency.**

In [17]:
# Add your code below
inspection_type_and_frequency_query = '''
SELECT
    type,
    COUNT(type) AS frequency
FROM
    inspections
GROUP BY
    type
ORDER BY frequency ASC
'''
inspection_type_and_frequency_result = pd.read_sql(inspection_type_and_frequency_query, conn)
inspection_type_and_frequency_result

type  frequency
0          Multi-agency Investigation          1
1   Administrative or Document Review          3
2                       Special Event          7
3                 Routine - Scheduled         74
4     Foodborne Illness Investigation         80
5     Complaint Reinspection/Followup        138
6               Structural Inspection        170
7           Non-inspection site visit        814
8                           Complaint       1625
9                    New Construction       1737
10                      New Ownership       1843
11              Reinspection/Followup       5283
12              Routine - Unscheduled      15568

**9. What is the average score given to restaurants based on the type of inspection? Based on the results, identify the types of inspections that are not scored (NULL) and remove those categories from the resultset. The 'average_score' should be rounded to one decimal. Sort the results in ascending order based on the average score. Hint: use the function `ROUND(score, 1)`**

In [18]:
# Add your code below
average_score_per_inspection_type_query = '''
SELECT
    type,
    ROUND(AVG(score), 1) AS average_score
FROM 
    inspections
GROUP BY 
    type
HAVING
    average_score != 'NaN'
ORDER BY
    average_score ASC
'''
average_score_per_inspection_type_result = pd.read_sql(average_score_per_inspection_type_query, conn)
average_score_per_inspection_type_result

type  average_score
0  Reinspection/Followup           85.0
1  Routine - Unscheduled           91.2
2          New Ownership          100.0

**10. Find the restaurant owners (owner_name) that own one or multiple restaurants in the city with the number of restaurants (num_restaurants) they own. Find the 10 owners who own the most restaurants and sort them by the number of restaurants they own in decreasing order.**

In [19]:
# Add your code below
owners_with_restaurant_numbers_query = '''
SELECT
    owner_name,
    COUNT(owner_name) AS num_restaurants
FROM 
    businesses
GROUP BY 
    owner_name
ORDER BY 
    num_restaurants DESC
LIMIT 
    10
'''
owners_with_restaurant_numbers_result = pd.read_sql(owners_with_restaurant_numbers_query, conn)
owners_with_restaurant_numbers_result

owner_name  num_restaurants
0                           VOLUME SERVICES AMERICA               94
1  Department of Children, Youth and their Families               64
2                                             SFUSD               44
3                             Volume Services, Inc.               41
4                       Children's Empowerment Inc.               37
5                      SAN FRANCISCO UNIFIED SCHOOL               34
6                                      WALGREEN CO.               17
7                                       Walgreen Co               15
8                         SMG Food and Beverage LLC               14
9                              STARBUCKS COFFEE CO.               13

## Part 3: Subqueries and joins

**11. From the businesses table, find all owners that own more than five restaurants. Then  find the 10 most popular locations for  restaurants (using postal_code) amongst the owners who have 5 restaurant or more. The final result should return the 10 most popular areas  (postal_code) and the frequency with which they appear in our group of owners who own five restaurants or more. The result should have two columns (postal_code and frequency) and should be presented in descending order of frequency.**

In [46]:
# Add your code below
sub_query_popular = '''
SELECT
    *
FROM
    businesses
GROUP BY
    owner_name
HAVING
    COUNT(*) > 5
'''

most_popular_post_codes_query = f'''
SELECT
    *
FROM 
    {sub_query_popular}
GROUP BY 
    postal_code
ORDER BY
    frequency DESC
LIMIT 10
'''
most_popular_post_codes_result = pd.read_sql(sub_query_popular, conn)
most_popular_post_codes_result

business_id                                          name  \
0           509                                   Cafe Bakery   
1         62767                             Lincoln Way Shell   
2         75277                         Bon Appetit @ Twitter   
3            81                           CLUB ED/BON APPETIT   
4         22177                                      Outtakes   
5         37166                        Luce Restaurant &  Bar   
6         23478                               Herz Playground   
7          6042             SFRP - HAMILTON RECREATION CENTER   
8         62674                  Chipotle Mexican Grill #1230   
9          6056                           RAD Westside Courts   
10        25151                    Westin St Francis - Caruso   
11        64726                            CVS Pharmacy #2852   
12        74485                           Guckenheimer@Nektar   
13        63646          SFDH - Harry Denton's Starlight Room   
14         5519                   AT&T - (J-3A) BEN & JERRY'S   
15         3562                     Mixed Greens Prep Kitchen   
16          394                  SF MARRIOTT HOTEL PANTRY 5/F   
17           24            OMNI S.F. Hotel - 2nd Floor Pantry   
18        29312                           Peet's Coffee & Tea   
19         3929           OPERA HOUSE - BASEMENT,LOWER LOUNGE   
20         5880                       Francisco Middle School   
21         3055                     SAFEWAY STORES, INC. #995   
22         5787                      LAWTON ELEMENTARY SCHOOL   
23        86159                                     Lobby Bar   
24         5801                       FLYNN ELEMENTARY SCHOOL   
25        19055            Moscone Center West - Bsmt Kitchen   
26         1085                        STARBUCKS COFFEE #9219   
27           66                                     STARBUCKS   
28         7464                                      Zelinsky   
29         3924                    San Francisco Soup Company   
30         5851                                ROOFTOP SCHOOL   
31         1368                              Starbucks Coffee   
32         1866                      Starbucks Coffee Company   
33        70181                           Target Store T-2766   
34         3977         AT&T - WILLIE MAYS PLAZA BAR [145087]   
35         1515  AT&T - Cognac Bar/ KITCHEN Room3232 [145080]   
36        86117                                   Jamba Juice   
37        69044                   AT&T PARK - NEW DERBY GRILL   
38        76355     AT&T PARK - Centerplate/Crazy Grab'z Cart   
39        74317                  AT&T PARK - DRAFT CART (P 8)   
40         7280                              WALGREENS #04180   
41         7311                                     Walgreens   
42         7327                                     WALGREENS   
43        62048                              Walgreens #09886   

                          address           city postal_code   latitude  \
0                 1365 Noriega St             SF       94122  37.754090   
1                 601 Lincoln Way             Sf       94122  37.766003   
2            1355 Market St 10/F.           None       94103        NaN   
3                    2350 TURK St           S.F.       94117  37.778468   
4                    2800 TURK St           S.F.       94117  37.778106   
5         888 Howard St 1st Floor             SF       94103  37.781664   
6             1701 Visitation Ave           None       94134        NaN   
7                   1900 GEARY St  San Francisco       94110        NaN   
8                232 O'Farrell St  San Francisco       94102  37.786387   
9                  2501 SUTTER St  San Francisco       94115        NaN   
10            335 Powell St Lobby  San Francisco       94102  37.787700   
11                  731 Market St             SF       94103  37.786844   
12     455 South Mission Bay Blvd           None       94158        NaN   
13       450 Powell St 21st 

**12. Now it might be interesting to look at some statistics. For all the restaurants in the "94103" post code let's calculate the minimum score (as "min_score"), average score (as "avg_score") and maximum S=score (as "max_score"). The average score should be rounded to one decimal and you should only consider restaurants that have undergone inspection (so the score is NOT NULL).**

In [ ]:
# Add your code below
# min_avg_max_score_query = ...
# min_avg_max_score_result = pd.read_sql(min_avg_max_score_query, conn)


**13. Now we can get a bit more serious and look at how many times restaurants in the "94103" post code  have committed health violations. We can then group them based on their risk category so we know which restaurants to avoid. The output should have two columns (risk_category, frequency) and be sorted in descending order by frequency**

In [ ]:
# Add your code below
# market_street_health_violations_query = ...
# market_street_health_violations_result = pd.read_sql(market_street_health_violations_query, conn)
